src/util/replace_symbol.cpp

```diff
@@ -89,33 +91,75 @@ bool replace_symbolt::replace(exprt &dest) const
   if(!have_to_replace(dest))
     return result;
 
-  if(dest.id()==ID_symbol)
+  if(dest.id()==ID_member)
+  {
+    member_exprt &me=to_member_expr(dest);
+
+    if(!replace(me.struct_op(), replace_with_const))
+      result=false;
+  }
+  else if(dest.id()==ID_index)
+  {
+    index_exprt &ie=to_index_expr(dest);
+
+    if(!replace(ie.array(), replace_with_const))
+      result=false;
+
+    if(!replace(ie.index()))
+      result=false;
+  }
+  else if(dest.id()==ID_address_of)
+  {
+    address_of_exprt &aoe=to_address_of_expr(dest);
+
+    if(!replace(aoe.object(), false))
+      result=false;
+  }
+  else if(dest.id()==ID_symbol)
   {
+    const symbol_exprt &s=to_symbol_expr(dest);
+
     expr_mapt::const_iterator it=
-      expr_map.find(dest.get(ID_identifier));
+      expr_map.find(s.get_identifier());
 
     if(it!=expr_map.end())
     {
-      dest=it->second;
+      const exprt &e=it->second;
+
+      if(!replace_with_const && e.is_constant())
+        return true;
+
+      dest=e;
+
       return false;
     }
   }
-
-  Forall_operands(it, dest)
-    if(!replace(*it))
-      result=false;
+  else
+  {
+    Forall_operands(it, dest)
+      if(!replace(*it))
```

Comment:
I don't think this is needed as it's only the "outermost" address-of, index, member, etc. that you need to catch.  The logic is similar to:

https://github.com/diffblue/cbmc/blob/master/src/analyses/ai.cpp#L83